### PR TITLE
Dynamic quantizatiion for Fx

### PR DESCRIPTION
**Pull request by DeepDriving**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49948 Dynamic quantizatiion for Fx**
* #49266 compare_model_outputs_fx API implementation
* #48951 compare_model_stub_fx API implementation

Dynamic quantization for FX
buck test @mode/dev caffe2/test:quantization_fx -- 'test_compare_weights_lstm_dynamic_fx'
buck test @mode/dev caffe2/test:quantization_fx -- 'test_compare_model_stub_linear_dynamic_fx'
buck test @mode/dev caffe2/test:quantization_fx -- 'test_compare_model_stub_lstm_dynamic_fx'
buck test @mode/dev caffe2/test:quantization_fx -- 'test_compare_model_outputs_linear_dynamic_fx'

Differential Revision: [D25726967](https://our.internmc.facebook.com/intern/diff/D25726967/)

### CODE REVIEW

File: test/quantization/test_numeric_suite_fx.py [hunk 1]
```diff
@@ -285,3 +382,32 @@ def compare_and_validate_results(float_model, q_model, data):
...
+    @override_qengines
+    def test_compare_model_outputs_linear_dynamic_fx(self):
+        r"""Compare the output of linear layer in dynamic quantized model and corresponding
+        output of linear layer in float model
+        """
+
+        def compare_and_validate_results(float_model, q_model, data):
+            act_compare_dict = compare_model_outputs_fx(float_model, q_model, data)
+            print(act_compare_dict)
+            expected_act_compare_dict_keys = {"stats"}
```
Comment (hx89) [inline test/quantization/test_numeric_suite_fx.py:395]: Expected key should have another one for fc1.

Comment (DeepDriving) [reply test/quantization/test_numeric_suite_fx.py]: Fixed.

File: test/quantization/test_numeric_suite_fx.py [hunk 2]
```diff
@@ -135,6 +136,34 @@ def compare_and_validate_results(float_model, q_model):
...
 
+    @override_qengines
+    def test_compare_weights_lstm_dynamic_fx(self):
+        r"""Compare the weights of float and dynamic quantized lstm layer"""
+
+        def compare_and_validate_results(float_model, q_model):
+            weight_dict = compare_weights_fx(
+                float_model.state_dict(), q_model.state_dict()
+            )
+            self.assertEqual(len(weight_dict), 1)
```
Comment (hx89) [inline test/quantization/test_numeric_suite_fx.py:147]: Could you add the check if the key in weight_dict is expected?

File: test/quantization/test_numeric_suite_fx.py [hunk 3]
```diff
@@ -214,6 +243,77 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
...
 
+    @override_qengines
+    def test_compare_model_stub_linear_dynamic_fx(self):
+        r"""Compare the output of dynamic quantized linear layer and its float shadow module"""
+
+        def compare_and_validate_results(float_model, q_model, module_swap_list, data):
+            ob_dict = compare_model_stub_fx(
+                float_model, q_model, module_swap_list, data
+            )
+            self.assertEqual(len(ob_dict), 1)
```
Comment (hx89) [inline test/quantization/test_numeric_suite_fx.py:254]: Could you add the check if the key in ob_dict is expected?

File: test/quantization/test_numeric_suite_fx.py [hunk 4]
```diff
@@ -214,6 +243,77 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
...
+    def test_compare_model_stub_lstm_dynamic_fx(self):
+        r"""Compare the output of dynamic quantized linear layer and its float shadow module"""
+
+        def compare_and_validate_results(
+            float_model, q_model, module_swap_list, input, hidden
+        ):
+            ob_dict = compare_model_stub_fx(
+                float_model, q_model, module_swap_list, input, hidden
+            )
+            self.assertEqual(len(ob_dict), 1)
```
Comment (hx89) [inline test/quantization/test_numeric_suite_fx.py:290]: Could you add the check if the key in ob_dict is expected?

File: test/quantization/test_numeric_suite_fx.py (hunk 2 above)
Comment (DeepDriving) [reply test/quantization/test_numeric_suite_fx.py:147]: Added in compare_model_stub PR

File: test/quantization/test_numeric_suite_fx.py (hunk 4 above)
Comment (DeepDriving) [reply test/quantization/test_numeric_suite_fx.py:290]: Added in compare_model_stub PR

**Review by raghuramank100 — COMMENTED** [review]
Is the change in this PR mainly the additional tests?
If so, the changes to _numeric_suite_fx dont seem to be used in the code.

**Pull request closed without merging**